datafusion/expr/src/expr_schema.rs

```diff
@@ -204,7 +205,12 @@ impl ExprSchemable for Expr {
                                     )
                                 )
                             })?;
-                        Ok(fun.return_type(&new_types, &nullability)?)
+                        let function_name = self.schema_name().to_string();
+                        let field_args =
+                            WindowUDFFieldArgs::new(&new_types, &function_name);
+
+                        udwf.field(field_args)
+                            .map(|field| field.data_type().clone())
```

Comment:
The data type of the Window UDF result is now derived from the `field` trait method. More verbose.